File: scripts/uat-crawl/evaluate-crawl-verdict.mjs
#!/usr/bin/env node
/**
 * Final certification verdict — collect-all then fail closed on material defects.
 * Must run after evidence collection steps (if: always()).
 */
import fs from "node:fs";
import path from "node:path";

const ROOT = path.resolve(import.meta.dirname, "../..");
const RUN_ID = process.env.GITHUB_RUN_ID || "local";
const RUN_TRANCHE = process.env.RUN_TRANCHE || "unknown";

function readJson(relativePath) {
  const filePath = path.join(ROOT, relativePath);
  if (!fs.existsSync(filePath)) return null;
  return JSON.parse(fs.readFileSync(filePath, "utf8"));
}

function readJsonl(relativePath) {
  const filePath = path.join(ROOT, relativePath);
  if (!fs.existsSync(filePath)) return [];
  return fs
    .readFileSync(filePath, "utf8")
    .trim()
    .split("\n")
    .filter(Boolean)
    .map((line) => JSON.parse(line));
}

const failures = [];
const warnings = [];

const outcomesPath = path.join(ROOT, "docs/uat-crawl/UAT_CRAWL_STEP_OUTCOMES.json");
const stepOutcomes = fs.existsSync(outcomesPath) ? readJson("docs/uat-crawl/UAT_CRAWL_STEP_OUTCOMES.json") : {};

if (process.env.UAT_DEPLOY_BLOCKED === "true") {
  failures.push("DEPLOY_BLOCKED: no trusted crawl target for current-main tranche");
}

for (const [step, outcome] of Object.entries(stepOutcomes)) {
  if (outcome === "failure" || outcome === "cancelled") {
    failures.push(`STEP_FAILED:${step}`);
  }
}
if (stepOutcomes.ai_uat === "failure") {
  failures.push("AI_UAT_SUITE_FAILED");
}
if (stepOutcomes.post_fix_483 === "failure") {
  failures.push("POST_FIX_483_SUITE_FAILED");
}

const blockersSummary = readJson("docs/uat-crawl/UAT_VERIFIED_BLOCKERS_SUMMARY.json");
if (blockersSummary && blockersSummary.denominatorReconciled === false) {
  failures.push("CENSUS_RECONCILIATION_FAILED");
}

const aiSummary = readJson("docs/uat-crawl/UAT_AI_UAT_SUMMARY.json");
if (aiSummary?.runId === RUN_ID) {
  if (stepOutcomes.ai_uat === "failure" && aiSummary.counts?.PASS > 0 && !aiSummary.priorRunArchived) {
    failures.push("AI_UAT_STALE_PASS: run failed but summary reports PASS from same run metadata conflict");
  }
  if (aiSummary.counts?.NOT_EXECUTED > 0 && stepOutcomes.ai_uat === "failure") {
    warnings.push("AI_UAT_NOT_EXECUTED_AFTER_FAILURE");
  }
} else if (stepOutcomes.ai_uat === "failure") {
  failures.push("AI_UAT_FAILED_WITHOUT_CURRENT_RUN_SUMMARY");
}

const ghaRun = readJson("docs/uat-crawl/UAT_GHA_RUN.json");
if (ghaRun?.deployBlocked === true) {
  failures.push("GHA_RUN_DEPLOY_BLOCKED");
}
if (ghaRun?.crawlTargetType && ghaRun.crawlTargetType !== "PUBLIC_PRODUCTION_ALIAS" && ghaRun.runTranche?.includes("watchdog")) {
  warnings.push(`CRAWL_TARGET_TYPE:${ghaRun.crawlTargetType}`);
}

const protectionRows = readJsonl("docs/uat-crawl/UAT_MANIFEST.jsonl")
  .concat(readJsonl("docs/uat-crawl/UAT_MANIFEST_AUTH.jsonl"))
  .filter((row) => row.blockClassification === "DEPLOYMENT_PROTECTION" || row.wallClassification === "DEPLOYMENT_PROTECTION");

if (protectionRows.length > 0) {
  failures.push(`DEPLOYMENT_PROTECTION:${protectionRows.map((r) => r.uatId).join(",")}`);
}

const rebaseline = readJson("docs/uat-crawl/UAT_REBASELINE_CURRENT_MAIN.json");
if (rebaseline?.runId === RUN_ID && rebaseline.deployProvenance?.includes("a619a7a2")) {
  failures.push("STALE_PROVENANCE_LABEL_IN_REBASELINE");
}

const provenance = readJson("docs/uat-crawl/UAT_DEPLOY_PROVENANCE.json");
if (provenance?.runId === RUN_ID && provenance.requiredShaStatus?.includes("a619a7a2")) {
  failures.push("STALE_PROVENANCE_LABEL_IN_DEPLOY_PROVENANCE");
}

const verdict = {
  generatedAt: new Date().toISOString(),
  runId: RUN_ID,
  runTranche: RUN_TRANCHE,
  passed: failures.length === 0,
  failures,
  warnings,
  stepOutcomes,
};

const outPath = path.join(ROOT, "docs/uat-crawl/UAT_CRAWL_CERTIFICATION_VERDICT.json");
fs.mkdirSync(path.dirname(outPath), { recursive: true });
fs.writeFileSync(outPath, `${JSON.stringify(verdict, null, 2)}\n`);

if (failures.length > 0) {
  for (const failure of failures) {
    console.error(`::error::UAT certification verdict FAILED — ${failure}`);
  }
  process.exitCode = 1;
} else {
  console.log(`UAT certification verdict PASSED (warnings=${warnings.length}).`);
}

File: scripts/uat-crawl/evaluate-crawl-verdict.py
#!/usr/bin/env python3
"""Final certification verdict — collect-all then fail closed on material defects."""
from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
DOCS = ROOT / "docs" / "uat-crawl"
RUN_ID = os.environ.get("GITHUB_RUN_ID", "local")
RUN_TRANCHE = os.environ.get("RUN_TRANCHE", "unknown")


def read_json(name: str) -> dict | None:
    path = DOCS / name
    if not path.is_file():
        return None
    return json.loads(path.read_text(encoding="utf-8"))


def read_jsonl(name: str) -> list[dict]:
    path = DOCS / name
    if not path.is_file():
        return []
    rows = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            rows.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return rows


def main() -> int:
    failures: list[str] = []
    warnings: list[str] = []
    step_outcomes = read_json("UAT_CRAWL_STEP_OUTCOMES.json") or {}

    if os.environ.get("UAT_DEPLOY_BLOCKED") == "true":
        failures.append("DEPLOY_BLOCKED: no trusted crawl target for current-main tranche")

    for step, outcome in step_outcomes.items():
        if outcome in {"failure", "cancelled"}:
            failures.append(f"STEP_FAILED:{step}")
    if step_outcomes.get("ai_uat") == "failure":
        failures.append("AI_UAT_SUITE_FAILED")
    if step_outcomes.get("post_fix_483") == "failure":
        failures.append("POST_FIX_483_SUITE_FAILED")

    blockers_summary = read_json("UAT_VERIFIED_BLOCKERS_SUMMARY.json")
    if blockers_summary and blockers_summary.get("denominatorReconciled") is False:
        failures.append("CENSUS_RECONCILIATION_FAILED")

    ai_summary = read_json("UAT_AI_UAT_SUMMARY.json")
    if ai_summary and ai_summary.get("runId") == RUN_ID:
        counts = ai_summary.get("counts") or {}
        if step_outcomes.get("ai_uat") == "failure" and counts.get("PASS", 0) > 0 and not ai_summary.get("priorRunArchived"):
            failures.append("AI_UAT_STALE_PASS: run failed but summary reports PASS from same run metadata conflict")
        if counts.get("NOT_EXECUTED", 0) > 0 and step_outcomes.get("ai_uat") == "failure":
            warnings.append("AI_UAT_NOT_EXECUTED_AFTER_FAILURE")
    elif step_outcomes.get("ai_uat") == "failure":
        failures.append("AI_UAT_FAILED_WITHOUT_CURRENT_RUN_SUMMARY")

    gha_run = read_json("UAT_GHA_RUN.json")
    if gha_run and gha_run.get("deployBlocked") is True:
        failures.append("GHA_RUN_DEPLOY_BLOCKED")
    if (
        gha_run
        and gha_run.get("crawlTargetType")
        and gha_run.get("crawlTargetType") != "PUBLIC_PRODUCTION_ALIAS"
        and "watchdog" in str(gha_run.get("runTranche", ""))
    ):
        warnings.append(f"CRAWL_TARGET_TYPE:{gha_run.get('crawlTargetType')}")

    protection_rows = [
        row
        for row in read_jsonl("UAT_MANIFEST.jsonl") + read_jsonl("UAT_MANIFEST_AUTH.jsonl")
        if row.get("blockClassification") == "DEPLOYMENT_PROTECTION"
        or row.get("wallClassification") == "DEPLOYMENT_PROTECTION"
    ]
    if protection_rows:
        failures.append("DEPLOYMENT_PROTECTION:" + ",".join(row.get("uatId", "?") for row in protection_rows))

    rebaseline = read_json("UAT_REBASELINE_CURRENT_MAIN.json")
    if rebaseline and rebaseline.get("runId") == RUN_ID and "a619a7a2" in str(rebaseline.get("deployProvenance", "")):
        failures.append("STALE_PROVENANCE_LABEL_IN_REBASELINE")

    provenance = read_json("UAT_DEPLOY_PROVENANCE.json")
    if provenance and provenance.get("runId") == RUN_ID and "a619a7a2" in str(provenance.get("requiredShaStatus", "")):
        failures.append("STALE_PROVENANCE_LABEL_IN_DEPLOY_PROVENANCE")

    verdict = {
        "generatedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "runId": RUN_ID,
        "runTranche": RUN_TRANCHE,
        "passed": len(failures) == 0,
        "failures": failures,
        "warnings": warnings,
        "stepOutcomes": step_outcomes,
    }
    out_path = DOCS / "UAT_CRAWL_CERTIFICATION_VERDICT.json"
    out_path.write_text(json.dumps(verdict, indent=2) + "\n", encoding="utf-8")

    if failures:
        for failure in failures:
            print(f"::error::UAT certification verdict FAILED — {failure}", file=sys.stderr)
        return 1

    print(f"UAT certification verdict PASSED (warnings={len(warnings)}).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
